### PR TITLE
Added extractVariables method to IOFactory

### LINKED ISSUE #2514: Add Variable Extraction Feature to IOFactory for Enhanced Spreadsheet Integration
## Overview
This pull request introduces a new method `extractVariables` to the `IOFactory` class in PHPWord. The primary goal of this enhancement is to enable the extraction of variables from a Word document, particularly those that are linked to spreadsheet cell coordinates. This functionality lays the groundwork for more advanced integrations between Word documents and spreadsheets, allowing for dynamic data retrieval and manipulation.

## Technical Details
The `extractVariables` method scans a Word document and extracts all instances of variables following a specific pattern (e.g., `${variable}`). This is particularly useful in scenarios where these variables correspond to cell coordinates in a spreadsheet. By extracting these variables, we can programmatically search for corresponding values in linked spreadsheets, thereby automating the process of updating Word documents with dynamic data from spreadsheets.

## Use Cases
- **Dynamic Document Generation:** Automate the generation of Word documents with real-time data from spreadsheets. For instance, generating financial reports, invoices, or personalized documents where key figures are drawn from a spreadsheet.
- **Template Processing:** Enhance template processing by allowing templates to specify spreadsheet data sources directly within the Word document.

## Implementation
The implementation involved modifying the `IOFactory` class to include the `extractVariables` method. This method leverages existing structures within PHPWord to navigate through a document's sections and elements, identifying and aggregating variables.

## Tests
Appropriate unit tests have been added to ensure the functionality works as expected and to prevent future regressions. These tests cover a range of scenarios, including documents with a variety of variable patterns and formats.

## Conclusion
This enhancement is a step towards deeper integration between Word documents and spreadsheets, unlocking new potentials in document automation and data-driven report generation. It aligns with PHPWord's goal of providing robust and flexible tools for word processing and document manipulation.

### PR DESCRIPTION
**Pull request by sibalonat**
## Overview
This pull request introduces a new method `extractVariables` to the `IOFactory` class in PHPWord. The primary goal of this enhancement is to enable the extraction of variables from a Word document, particularly those that are linked to spreadsheet cell coordinates. This functionality lays the groundwork for more advanced integrations between Word documents and spreadsheets, allowing for dynamic data retrieval and manipulation.

## Technical Details
The `extractVariables` method scans a Word document and extracts all instances of variables following a specific pattern (e.g., `${variable}`). This is particularly useful in scenarios where these variables correspond to cell coordinates in a spreadsheet. By extracting these variables, we can programmatically search for corresponding values in linked spreadsheets, thereby automating the process of updating Word documents with dynamic data from spreadsheets.

## Use Cases
- **Dynamic Document Generation:** Automate the generation of Word documents with real-time data from spreadsheets. For instance, generating financial reports, invoices, or personalized documents where key figures are drawn from a spreadsheet.
- **Template Processing:** Enhance template processing by allowing templates to specify spreadsheet data sources directly within the Word document.

## Implementation
The implementation involved modifying the `IOFactory` class to include the `extractVariables` method. This method leverages existing structures within PHPWord to navigate through a document's sections and elements, identifying and aggregating variables.

## Tests
Appropriate unit tests have been added to ensure the functionality works as expected and to prevent future regressions. These tests cover a range of scenarios, including documents with a variety of variable patterns and formats.

## Conclusion
This enhancement is a step towards deeper integration between Word documents and spreadsheets, unlocking new potentials in document automation and data-driven report generation. It aligns with PHPWord's goal of providing robust and flexible tools for word processing and document manipulation.

Fixes #2514

### Checklist:

- [X ] I have run `composer run-script check --timeout=0` and no errors were reported
- [ X] The new code is covered by unit tests (check build/coverage for coverage report)
- [ -] I have updated the documentation to describe the changes

### CODE REVIEW

File: src/PhpWord/IOFactory.php [hunk 1]
```diff
@@ -89,6 +90,44 @@ public static function load($filename, $readerName = 'Word2007')
...
 
+    /**
+     * Loads PhpWord ${variable} from file.
+     *
+     * @param string $filename The name of the file
+     * @param string $readerName
+     *
+     * @return array The extracted variables
+     */
+    public static function extractVariables(string $filename, $readerName = 'Word2007')
```
Comment (Progi1984) [inline src/PhpWord/IOFactory.php:101]:
```suggestion
    public static function extractVariables(string $filename, string $readerName = 'Word2007'): array
```

File: samples/Sample_44_ExtractVariablesFromReaderWord2007.php [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+include_once 'Sample_Header.php';
+
+// Read contents
+$name = basename(__FILE__, '.php');
+$source = __DIR__ . "/resources/{$name}.docx";
+echo date('H:i:s'), " Reading contents from `{$source}`", EOL;
+$phpWord = \PhpOffice\PhpWord\IOFactory::extractVariables($source);
```
Comment (Progi1984) [inline samples/Sample_44_ExtractVariablesFromReaderWord2007.php:9]:
```suggestion
$variables = \PhpOffice\PhpWord\IOFactory::extractVariables($source);

var_dump($variables);
```

**Review by Progi1984 — COMMENTED** [review]
@sibalonat Great improvement which improve the Developer Experience.

Could you fix my comments and update the changelog 2.0.0.md file, please ?

**Comment by sibalonat** [comment]
I would gladly do so @Progi1984. Thanks for the comments and ill let you know as soon as it's done.

File: samples/Sample_44_ExtractVariablesFromReaderWord2007.php (hunk 2 above)
Comment (sibalonat) [reply samples/Sample_44_ExtractVariablesFromReaderWord2007.php]: on it

File: docs/changes/2.x/2.0.0.md [hunk 3]
```diff
@@ -4,6 +4,8 @@
 
 ## Enhancements
 
+- Word2007 Writer : Added extractVariables method to extract variables from word [@sibalonat](https://github.com/sibalonat) in GH-2515
```
Comment (Progi1984) [inline docs/changes/2.x/2.0.0.md:7]:
```suggestion
- IOFactory : Added extractVariables method to extract variables from a document [@sibalonat](https://github.com/sibalonat) in [#2515](https://github.com/PHPOffice/PHPWord/pull/2515)
```

File: src/PhpWord/IOFactory.php [hunk 4]
```diff
@@ -12,11 +12,12 @@
  *
  * @see         https://github.com/PHPOffice/PHPWord
  *
- * @license     http://www.gnu.org/licenses/lgpl.txt LGPL version 3
- */
+ * @license     http://www.gnu.org/licenses/lgpl.txt LGPL version 3 */
```
Comment (Progi1984) [inline src/PhpWord/IOFactory.php:15]:
```suggestion
 * @license     http://www.gnu.org/licenses/lgpl.txt LGPL version 3
 */
```

File: src/PhpWord/IOFactory.php [hunk 5]
```diff
@@ -89,6 +90,44 @@ public static function load($filename, $readerName = 'Word2007')
...
 
+    /**
+     * Loads PhpWord ${variable} from file.
+     *
+     * @param string $filename The name of the file
+     * @param string $readerName
+     *
+     * @return array The extracted variables
+     */
+    public static function extractVariables(string $filename, $readerName = 'Word2007'): array
```
Comment (Progi1984) [inline src/PhpWord/IOFactory.php:101]:
```suggestion
    public static function extractVariables(string $filename, string $readerName = 'Word2007'): array
```

**Review by Progi1984 — COMMENTED** [review]
Some last feedbacks

**Comment by sibalonat** [comment]
Ill add them now, thanks again, much appreciated :) @Progi1984

**Review by Progi1984 — APPROVED** [review]

**Pull request merged**